nixos/lib/test-driver/test_driver/machine.py

```diff
@@ -1083,15 +1094,18 @@ def create_socket(path: Path) -> socket.socket:
 
         monitor_socket = create_socket(clear(self.monitor_path))
         shell_socket = create_socket(clear(self.shell_path))
+        self.qmp_client = QMPClient(self.name)
```

Comment:
Since everything in nixos is using sync python, can we properly use this async qmp client?

Otherwise this might be also good enough?

```
class QmpSession:
    def __init__(self, sock: socket.socket) -> None:
        self.sock = sock
        self.pending_events: Queue[Dict[str, Any]] = Queue()
        self.reader = sock.makefile("r")
        self.writer = sock.makefile("w")
        hello = self._result()
        assert "QMP" in hello, f"Unexpected result: {hello}"
        self.send("qmp_capabilities")

    def _readmsg(self) -> Dict[str, Any]:
        line = self.reader.readline()
        return json.loads(line)

    def _raise_unexpected_msg(self, msg: Dict[str, Any]) -> None:
        m = json.dumps(msg, sort_keys=True, indent=4)
        raise RuntimeError(f"Got unexpected qmp response: {m}")

    def _result(self) -> Dict[str, Any]:
        while True:
            # QMP is in the handshake
            res = self._readmsg()
            if "return" in res or "QMP" in res:
                return res
            elif "event" in res:
                self.pending_events.put(res)
                continue
            else:
                self._raise_unexpected_msg(res)

    def events(self) -> Iterator[Dict[str, Any]]:
        while not self.pending_events.empty():
            yield self.pending_events.get()

        res = self._readmsg()

        if "event" not in res:
            self._raise_unexpected_msg(res)
        yield res

    def send(self, cmd: str, args: Dict[str, str] = {}) -> Dict[str, str]:
        data: Dict[str, Any] = dict(execute=cmd)
        if args != {}:
            data["arguments"] = args

        json.dump(data, self.writer)
        self.writer.write("\n")
        self.writer.flush()
        return self._result()


@contextmanager
def connect_qmp(path: Path) -> Iterator[QmpSession]:
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.connect(str(path))

    try:
        yield QmpSession(sock)
    finally:
        sock.close()
```

Reply:
Right, that makes sense

Reply:
@Mic92 Meh… I think we could have used the QMP async API by putting it in a second thread for the loop and still schedule stuff on that event loop. 

With this sync version, there's no automatic queueing of events, you have to call `events` yourself and accumulate the event one by one, while it's not obvious to me what kind of UX we want in the test driver with such an API (5 AM here…), problem starts when 

```
> machine # Cold-resetting the system now!
       > (finished: cleanup, in 0.03 seconds)
       > Traceback (most recent call last):
       >   File "/nix/store/06cfpx6c775i33nx8x2p5a6l4wqr537n-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
       >     sys.exit(main())
       >   File "/nix/store/06cfpx6c775i33nx8x2p5a6l4wqr537n-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/__init__.py", line 118, in main
       >     driver.run_tests()
       >   File "/nix/store/06cfpx6c775i33nx8x2p5a6l4wqr537n-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 147, in run_tests
       >     self.test_script()
       >   File "/nix/store/06cfpx6c775i33nx8x2p5a6l4wqr537n-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 143, in test_script
       >     exec(self.tests, symbols, None)
       >   File "<string>", line 8, in <module>
       >   File "/nix/store/06cfpx6c775i33nx8x2p5a6l4wqr537n-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/qmp.py", line 51, in events
       >     res = self._readmsg()
       >   File "/nix/store/06cfpx6c775i33nx8x2p5a6l4wqr537n-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/qmp.py", line 29, in _readmsg
       >     return json.loads(line)
       >   File "/nix/store/pzf6dnxg8gf04xazzjdwarm7s03cbrgz-python3-3.10.12/lib/python3.10/json/__init__.py", line 346, in loads
       >     return _default_decoder.decode(s)
       >   File "/nix/store/pzf6dnxg8gf04xazzjdwarm7s03cbrgz-python3-3.10.12/lib/python3.10/json/decoder.py", line 337, in decode
       >     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
       >   File "/nix/store/pzf6dnxg8gf04xazzjdwarm7s03cbrgz-python3-3.10.12/lib/python3.10/json/decoder.py", line 355, in raw_decode
       >     raise JSONDecodeError("Expecting value", s, err.value) from None
       > json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

happens, which seems to indicate that the reader is not blocking for data and more annoyingly, we probably need a non-blocking version of `events()` and thus `readmsg` because we would like to measure things like timeouts.

Or `events` needs a timeout parameter.

Either case, this code requires more work to be usable in the test driver.